File: MobMuPlatEditor-OSXLionAndUp/MobMuPlatEditor/Controls/MMPKnob.m
//
//  MMPKnob.m
//  MobMuPlatEd1
//
//  Created by Daniel Iglesia on 12/30/12.
//  Copyright (c) 2012 Daniel Iglesia. All rights reserved.
//

#import "MMPKnob.h"
#define ROTATION_PAD_RAD .7 //angle of the non-scrollable part at the bottom of the knob
#define EXTRA_RADIUS 10 //how many pixels from the edge of the knob are the tick marks
#define TICK_DIM 10 //size of tick marks

@implementation MMPKnob

int osxMinorVersion=-1;

//OSX 10.7 and 10.8 handle angles, rotation differently, so I need to know what version we are running.
+(int)osxMinorVersion{
    
    if(osxMinorVersion!=-1) return osxMinorVersion;//if we have already found it, send it out
    //otherwise, find it!
    else{
        osxMinorVersion=8;//assume 8;
   
        NSString* versionString = [[NSDictionary dictionaryWithContentsOfFile:@"/System/Library/CoreServices/SystemVersion.plist"] objectForKey:@"ProductVersion"];
        NSArray* versions = [versionString componentsSeparatedByString:@"."];
        //check( versions.count >= 2 );
        
        if ( versions.count >= 2 ) {
            osxMinorVersion = (int)[versions[1] integerValue];
            NSLog(@"OSX minor version:%d", osxMinorVersion);
            return osxMinorVersion;
        }
        else return osxMinorVersion;
    }
}

- (id)initWithFrame:(NSRect)frame{
    
    self = [super initWithFrame:frame];
    if (self) {
        self.address=@"/myKnob";
        
        [self setRange:1];//default
        
        knobView = [[NSView alloc]init];//create, don't set frame until setFrame
        [knobView setWantsLayer:YES];
        knobView.layer.backgroundColor=self.color.CGColor;
        [self addSubview:knobView];
        
        indicatorView=[[NSView alloc]init ];//create, don't set frame until setFrame
        [indicatorView setWantsLayer:YES];
    	indicatorView.layer.cornerRadius=3;
        [self addSubview:indicatorView];
        
		[self setColor:self.color];
        //[self setFrame:frame];
		[self addHandles];
      [self resizeSubviewsWithOldSize:self.frame.size];
        
        _value=0;
        [self updateIndicator];
        
    }
    
    return self;
}


-(void)setIndicatorColorUndoable:(NSColor*)inColor{
    [[self undoManager] registerUndoWithTarget:self selector:@selector(setIndicatorColorUndoable:) object:self.indicatorColor];
    [self setIndicatorColor:inColor];
}

-(void)setIndicatorColor:(NSColor*)inColor{
    _indicatorColor = inColor;
    indicatorView.layer.backgroundColor=inColor.CGColor;
}

-(void)hackRefresh{
    [super hackRefresh];
    knobView.layer.cornerRadius=radius;
    indicatorView.layer.cornerRadius=3;
    indicatorView.layer.backgroundColor=_indicatorColor.CGColor;
    for(NSView* dot in tickViewArray)dot.layer.cornerRadius=TICK_DIM/2;
    [self updateIndicator];
}

- (void)resizeSubviewsWithOldSize:(NSSize)oldBoundsSize{ //TODO use layoutSubviews.
  [super resizeSubviewsWithOldSize:oldBoundsSize];


  dim = self.frame.size.width-((EXTRA_RADIUS+TICK_DIM)*2);//diameter of circle

  //rounded up to nearest int - for corner radius
  radius = (float)(int)(dim/2+.5);

  [knobView setFrame:CGRectMake(EXTRA_RADIUS+TICK_DIM, EXTRA_RADIUS+TICK_DIM, dim, dim)];
  knobView.layer.cornerRadius = radius;

  indicatorDim=dim/2+2;
  indicatorThickness = dim/8;

  centerPoint=CGPointMake(dim/2+EXTRA_RADIUS+TICK_DIM, dim/2+EXTRA_RADIUS+TICK_DIM);
  [indicatorView setFrame:CGRectMake(centerPoint.x-indicatorThickness/2,centerPoint.y-indicatorThickness/2, indicatorThickness, indicatorDim)];
  [self updateIndicator];

  //tickmarks
  for(NSView* dot in tickViewArray){
    float angle= /*M_PI/2+M_PI-*/((float)[tickViewArray indexOfObject:dot]/(tickViewArray.count - 1) * (M_PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD+M_PI/2);/**/
    float xPos=(dim/2+EXTRA_RADIUS+TICK_DIM/2)*cos(angle);
    float yPos=(dim/2+EXTRA_RADIUS+TICK_DIM/2)*sin(angle);
    [dot setFrame:CGRectMake(centerPoint.x+xPos-(TICK_DIM/2),centerPoint.y+yPos-(TICK_DIM/2), TICK_DIM, TICK_DIM)];
  }
}

-(void)setFrame:(NSRect)frameRect{
    //unlike other MMPControls, we always keep this frame square
    
    //get larger of the dragged width and height to make a square
    int newDim = (frameRect.size.width>frameRect.size.height) ? frameRect.size.width : frameRect.size.height;
    CGRect squareFrame = CGRectMake(frameRect.origin.x, frameRect.origin.y, newDim, newDim);
    [super setFrame:squareFrame];
}

-(void)setColor:(NSColor *)color{
    [super setColor:color];
    knobView.layer.backgroundColor=color.CGColor;
    for(NSView* dot in tickViewArray)dot.layer.backgroundColor = color.CGColor;
}

-(void)setRangeObjectUndoable:(NSNumber*)inRangeObject{
    [[self undoManager] registerUndoWithTarget:self selector:@selector(setRangeObjectUndoable:) object:[NSNumber numberWithInt:self.range] ];
    [self setRange:[inRangeObject intValue] ];
}

-(void)setLegacyRange:(int)range {
  // old mode, default range is 2, which is range 0 to 1 float. translate that to new range of "1".
  if (range == 2) range = 1;
  [self setRange:range];
}

-(void)setRange:(int)inRange{
    
    _range=inRange;
    if(_range<1)_range=1;
    if(_range>1000)_range=1000;
    
    //remove and remake the tickViewArray
    for(NSView* dot in tickViewArray){
        [dot removeFromSuperview];
    }
    tickViewArray = [[NSMutableArray alloc]init];

    NSUInteger tickRange = _range == 1 ? 2 : _range;
    for(int i=0;i<tickRange;i++){
        NSView* dot = [[NSView alloc]init];
        [dot setWantsLayer:YES];
        dot.layer.backgroundColor=self.color.CGColor;
        dot.layer.cornerRadius=TICK_DIM/2;
        //printf("\n%.2f, %.2f", dot.center.x, dot.center.y);
        [self addSubview:dot];
        [tickViewArray addObject:dot];
    }
  //[self setNeedsLayout:YES];
  [self resizeSubviewsWithOldSize:self.frame.size];
  [self bringUpHandle]; // keep new view from being on top of edithandle, which can cause undo issues.
}

-(void)setValue:(float)inVal{
    if(inVal!=_value){//only on change

        if(_range==1){//if range is 1, clip value to 0.-1.
            if(inVal>1)inVal=1;
            if(inVal<0)inVal=0;
        }
        else{
            if(fmod(inVal, 1.0)!=0.0)inVal=(float)(int)inVal;//round inVal down to nearest integer
            if (inVal>=_range) {
                inVal=(float)(_range-1);//and clip if necessary
            }
        }

        _value=inVal;
        
        [self updateIndicator];
    }
}

//send out OSC message with value
-(void)sendValue{
    NSMutableArray* formattedMessageArray = [[NSMutableArray alloc]init];
    [formattedMessageArray addObject:self.address];
    if(_range>1){
        [formattedMessageArray addObject:[NSNumber numberWithInt:(int)self.value]];
    }
    else{
        [formattedMessageArray addObject:[NSNumber numberWithFloat:self.value]];
    }
    [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
}

// we have a new value, set the knob's indicator
//plenty of ugly math in this one...
-(void)updateIndicator{
	float newRad=0; //the angle of the indicatorView's rotation
    CGPoint newOrigin;

    if(_range==1){
        if ([MMPKnob osxMinorVersion]>=8) newRad= M_PI-((1.0-_value)*(M_PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD+M_PI/2);
        else newRad= M_PI-(_value*(M_PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD+M_PI/2);
    }
    
    else if (_range>1){
        if ([MMPKnob osxMinorVersion]>=8) newRad= M_PI-(((_range-_value-1)/(_range-1))*(M_PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD+M_PI/2);
        else newRad= M_PI-((_value/(_range-1))*(M_PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD+M_PI/2);
    }
    
    //set origin
    if ([MMPKnob osxMinorVersion]>=8) newOrigin = CGPointMake(centerPoint.x+sin(newRad)*(-indicatorThickness/2), centerPoint.y-cos(newRad)*(-indicatorThickness/2));
    else newOrigin = CGPointMake(centerPoint.x+sin(newRad)*(-indicatorThickness/2), centerPoint.y+cos(newRad)*(-indicatorThickness/2));
   
    [indicatorView setFrameOrigin:newOrigin];
    
    float newRad2 = (newRad-M_PI/2)/6.282*360.0;//offset and convert to degrees
        [indicatorView setFrameRotation:newRad2];
}


-(void)mouseDown:(NSEvent *)event{
    [super mouseDown:event];
    
   if(![self.editingDelegate isEditing] && self.enabled){
       knobView.layer.backgroundColor = self.highlightColor.CGColor;
     for(NSView* dot in tickViewArray)dot.layer.backgroundColor = self.highlightColor.CGColor;
       [self mouseDragged:event];
    }
}

//compute the value based on where I am touching
-(void)mouseDragged:(NSEvent *)event{	
	[super mouseDragged:event];
    if(![self.editingDelegate isEditing] && self.enabled){
        
        CGPoint point = [self convertPoint:[event locationInWindow] fromView:nil];//[[touches anyObject] locationInView:self];
    
        float touchX = point.x-centerPoint.x;
        float touchY = point.y-centerPoint.y;
        double theta = atan2(touchY, touchX);//raw theta =0 starting at 3 o'clock and going potive counterclockwise
    
        double updatedTheta = fmod( theta+M_PI/2+M_PI, (M_PI*2) );//theta =0 at 6pm going positive clockwise
   
        if(_range==1){
            if(updatedTheta<ROTATION_PAD_RAD)[self setValue:0];
            else if(updatedTheta>(M_PI*2-ROTATION_PAD_RAD)) [self setValue:1];
            else [self setValue:(updatedTheta-ROTATION_PAD_RAD)/(M_PI*2-2*ROTATION_PAD_RAD) ];
        
        }
        else if (_range>1){
            if(updatedTheta<ROTATION_PAD_RAD)[self setValue:0];
            else if(updatedTheta>(M_PI*2-ROTATION_PAD_RAD)) [self setValue:_range-1];
            else [self setValue:(float) (  (int)((updatedTheta-ROTATION_PAD_RAD)/(M_PI*2-2*ROTATION_PAD_RAD)*(_range-1)+.5)  ) ];//round to nearest tick!
        }
        [self sendValue];
    }
}

-(void)mouseUp:(NSEvent *)event{
    [super mouseUp:event];
    if(![self.editingDelegate isEditing] && self.enabled){
        knobView.layer.backgroundColor = self.color.CGColor;
      for(NSView* dot in tickViewArray)dot.layer.backgroundColor = self.color.CGColor;

    }
}

//receive messages from PureData (via [send toGUI], routed through the PdWrapper.pd patch), routed from Document via the address to this object
-(void)receiveList:(NSArray *)inArray{
  [super receiveList:inArray];
    BOOL sendVal=YES;
    //if message preceded by "set", then set "sendVal" flag to NO, and strip off set and make new messages array without it
    if ([inArray count]>0 && [[inArray objectAtIndex:0] isKindOfClass:[NSString class]] && [[inArray objectAtIndex:0] isEqualToString:@"set"]){
        NSRange newRange = (NSRange){1, [inArray count]-1};
        inArray = [inArray subarrayWithRange: newRange];
        //printf("\nset!");
        sendVal=NO;
    }
    //set new value, send it out 
    if ([inArray count]>0 && [[inArray objectAtIndex:0] isKindOfClass:[NSNumber class]]){
        [self setValue:[(NSNumber*)[inArray objectAtIndex:0] floatValue]];
        if(sendVal)[self sendValue];
    }
}

//coder for copy/paste

- (void)encodeWithCoder:(NSCoder *)coder {
    [super encodeWithCoder:coder];
    [coder encodeObject:self.indicatorColor forKey:@"indicatorColor"];
	[coder encodeInt:self.range forKey:@"range"];
    
}

- (id)initWithCoder:(NSCoder *)coder {
    
    if(self=[super initWithCoder:coder]){
        [self setIndicatorColor:[coder decodeObjectForKey:@"indicatorColor"]];
        [self setRange:[coder decodeIntForKey:@"range"]];
    }
    return self;
}


@end

File: MobMuPlatEditor-Java/src/com/iglesiaintermedia/MobMuPlatEditor/controls/MMPKnob.java
package com.iglesiaintermedia.MobMuPlatEditor.controls;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.RenderingHints;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.util.ArrayList;

import javax.swing.JPanel;


public class MMPKnob extends MMPControl implements MouseListener, MouseMotionListener{

	final static float ROTATION_PAD_RAD=.7f;
	final static int EXTRA_RADIUS=10;
	final static int TICK_DIM=10;
	
	private RoundedPanel knobPanel;
	private IndicatorPanel indicatorPanel;
	
	private int range;
	private ArrayList<RoundedPanel> tickViewArray;
	float value;
	private Color _indicatorColor, _indicatorDisabledColor;
	
	private float dim;
	private float radius;
	private float indicatorDim;
	private float indicatorThickness;
	private Point centerPoint;
	
	//copy constructor
	public MMPKnob(MMPKnob otherKnob){
		this(otherKnob.getBounds());//normal constructor
		this.setColor(otherKnob.getColor());
		this.setHighlightColor(otherKnob.getHighlightColor());
		this.address=otherKnob.address;
		this.setRange(otherKnob.range);
		this.setIndicatorColor(otherKnob.getIndicatorColor());
	}
	
	public MMPKnob(Rectangle frame){
		super();
		address="/myKnob";
		
		indicatorPanel = new IndicatorPanel();
		this.add(indicatorPanel);
		
		knobPanel = new RoundedPanel();	
		this.add(knobPanel);
		
		setRange(1);
		this.addMouseListener(this);
		this.addMouseMotionListener(this);
		this.setIndicatorColor(Color.WHITE);
		this.setColor(this.getColor());
		
		this.setBounds(frame);
		
		updateIndicator();
		
	}
	
	public Color getIndicatorColor() {
		return _indicatorColor;
	}
	
	public void setIndicatorColor(Color inColor){
		_indicatorColor=inColor;
		_indicatorDisabledColor = new Color(inColor.getRed(), inColor.getGreen(), inColor.getBlue(), (int)(inColor.getAlpha() * .2)); 
		indicatorPanel.setBackground(inColor);
	}
	
	public void setBounds(Rectangle frame){
		
		int newDim = (frame.width>frame.height) ? frame.width : frame.height;
		Rectangle squareFrame = new Rectangle(frame.x, frame.y, newDim, newDim);
		super.setBounds(squareFrame);
		dim = squareFrame.width-((EXTRA_RADIUS+TICK_DIM)*2);//diameter of circle 
		
		 //rounded up to nearest int - for corner radius
	    radius = (float)(int)(dim/2+.5);
	    knobPanel.setBounds(new Rectangle(EXTRA_RADIUS+TICK_DIM, EXTRA_RADIUS+TICK_DIM, (int)dim, (int)dim));
	    knobPanel.setCornerRadius((int)radius);
	   
	    indicatorDim=dim/2+2;
	    indicatorThickness = dim/8;
	    
	    centerPoint=new Point((int)(dim/2+EXTRA_RADIUS+TICK_DIM), (int)(dim/2+EXTRA_RADIUS+TICK_DIM));
	    //indicatorPanel.setBounds(new Rectangle ((int)(centerPoint.x-indicatorThickness/2), (int)(centerPoint.y-indicatorThickness/2), (int)indicatorThickness, (int)indicatorDim));
	    indicatorPanel.setBounds(0,0,getWidth(),getHeight());
	    updateIndicator();
	    
	    
	    for(RoundedPanel dot : tickViewArray){
	    	float angle= (float)((float)tickViewArray.indexOf(dot)/(tickViewArray.size()-1)* (Math.PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD+Math.PI/2);
	    	float xPos=(float) ( (dim/2+EXTRA_RADIUS+TICK_DIM/2)*Math.cos(angle) );
	    	float yPos=(float) ( (dim/2+EXTRA_RADIUS+TICK_DIM/2)*Math.sin(angle) );
	    	dot.setBounds(new Rectangle((int)(centerPoint.x+xPos-(TICK_DIM/2) ), (int)(centerPoint.y+yPos-(TICK_DIM/2) ), TICK_DIM, TICK_DIM) );
	    }
	}
	    
	  
	public void setColor(Color newColor){
		super.setColor(newColor);
		knobPanel.setBackground(newColor);
		for(JPanel tick: tickViewArray)tick.setBackground(newColor);
	}
	
	public void setLegacyRange(int range) {
		// translate old range value into new
		if(range == 2) range = 1;
		setRange(range);
	}
	
	public void setRange(int inRange){
	    range=inRange;
	    if(range<1)range=1;
	    if(range>1000)range=1000;
	    
	    if(tickViewArray!=null )
	    	for(JPanel tick:tickViewArray)tick.getParent().remove(tick);
	    
	    tickViewArray = new ArrayList<RoundedPanel>();
	    
	  int effectiveRange = range == 1 ? 2 : range; 
      for(int i=0;i<effectiveRange;i++){
    	  RoundedPanel tick = new RoundedPanel();
	      tick.setBackground(this.getColor());
	      tick.setCornerRadius((int)(TICK_DIM/2));
	      tickViewArray.add(tick);
	      add(tick);            
	   }
	    
	   this.setBounds(this.getBounds());
	   this.repaint();
	}

	public int getRange() {
		return range;
	}
	
	void setValue(float inVal){
		if(inVal!=value){
			if(range==1){//clip 0.-1.
		        if(inVal>1)inVal=1;
		        if(inVal<0)inVal=0;
		    }
			else{
		        if((inVal % 1.0)!=0.0)inVal=(float)(int)inVal;//round down to integer
		        if (inVal>=range) {
		            inVal=(float)(range-1);//clip if necessary
		        }
		    }
		    value=inVal;
		    //System.out.println(value);
		    updateIndicator();
			
		}
	}
	
	//send OSC message out
	public void sendValue(){
	   Object[] args = new Object[]{new Float(value)};
		editingDelegate.sendMessage(address, args);
	}

	
	void updateIndicator(){
		double newRad=0;
		if(range==1)
	        newRad= value*(Math.PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD;
		
		 else if (range>1)
		     newRad=(value/(range-1))*(Math.PI*2-ROTATION_PAD_RAD*2)+ROTATION_PAD_RAD;
		
	    indicatorPanel.setRotation(newRad);
	  
	}
	
	
	
	public void mousePressed(MouseEvent e) {
		//System.out.print(e.getClickCount() + " click(s)");
		super.mousePressed(e);
			
	    if(!editingDelegate.isEditing()){
	      knobPanel.setBackground(this.getHighlightColor());
	      // TODO handle enable/disable changes with color/highlightcolor changes
	      for(JPanel tick: tickViewArray)tick.setBackground(this.getHighlightColor());
	      mouseDragged(e);
	    }
	}

	@Override
	public void mouseClicked(MouseEvent e) {
		/*super.mouseClicked(e);
		if(!editingDelegate.isEditing()){
			thumbPanel.setLocation(e.getX(), e.getY());
		}*/
		
	}

	@Override
	public void mouseEntered(MouseEvent e) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void mouseExited(MouseEvent e) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void mouseReleased(MouseEvent e) {
		super.mouseReleased(e);
	    if(!editingDelegate.isEditing()){
	    	 knobPanel.setBackground(getColor());
	    	 for(JPanel tick: tickViewArray)tick.setBackground(getColor());
		       // indicatorColor.setBackground(indicatorColor);
		       
	    }
		
	}

	@Override
	public void mouseDragged(MouseEvent e) {
		super.mouseDragged(e);
		
		if(!editingDelegate.isEditing()){
			
	        float touchX = e.getX()-centerPoint.x;
	        float touchY = e.getY()-centerPoint.y;
	        double theta = Math.atan2(touchY, touchX);//raw theta (-pi to pi) =0 starting at 3 o'clock and going potive clockwise
	        
	        double updatedTheta = (theta-Math.PI/2+(Math.PI*2)) % (Math.PI*2) ;//theta (0 to 2pi) =0 at 6pm going positive clockwise
	        //System.out.print("\ntheta "+theta+" update "+updatedTheta);
	        if(range==1){
	            if(updatedTheta<ROTATION_PAD_RAD)setValue(0);
	            else if(updatedTheta>(Math.PI*2-ROTATION_PAD_RAD)) setValue(1);
	            else  setValue( (float)( (updatedTheta-ROTATION_PAD_RAD)/(Math.PI*2-2*ROTATION_PAD_RAD) ));
	        
	        }
	        else if (range>1){
	            if(updatedTheta<ROTATION_PAD_RAD)setValue(0);
	            else if(updatedTheta>(Math.PI*2-ROTATION_PAD_RAD)) setValue(range-1);
	            else setValue( (float) (  (int)((updatedTheta-ROTATION_PAD_RAD)/(Math.PI*2-2*ROTATION_PAD_RAD)*(range-1)+.5)  ) );//round to nearest tick!
	        }
	         sendValue();
		}
			
			
		
	}

	@Override
	public void mouseMoved(MouseEvent arg0) {
		// TODO Auto-generated method stub
		
	}
	
	//receive messages from PureData (via [send toGUI], routed through the PdWrapper.pd patch), routed from Document via the address to this object
	public void receiveList(ArrayList<Object> messageArray){
		super.receiveList(messageArray);
		boolean sendVal  = true;
		//if message preceded by "set", then set "sendVal" flag to NO, and strip off set and make new messages array without it
	    if (messageArray.size()>0 && (messageArray.get(0) instanceof String) && messageArray.get(0).equals("set") ){
	    	messageArray = new ArrayList<Object>(messageArray.subList(1, messageArray.size() ) );
	    	sendVal=false;
	    }
	    //set new value
	    if (messageArray.size()>0 && (messageArray.get(0) instanceof Integer) ){
	        setValue( ((Integer)(messageArray.get(0))).intValue() );
	        if(sendVal)sendValue();
	    }
	    if (messageArray.size()>0 && (messageArray.get(0) instanceof Float) ){
	        setValue( ((Float)(messageArray.get(0))).floatValue() );
	        if(sendVal)sendValue();
	    }

	}
	
	public void setEnabled(boolean enabled){
		super.setEnabled(enabled);
		Color c = enabled ? getColor() : getDisabledColor();
		knobPanel.setBackground(c);
		for(JPanel tick: tickViewArray)tick.setBackground(c);
		indicatorPanel.setBackground(enabled ? _indicatorColor : _indicatorDisabledColor);
	}


	class IndicatorPanel extends JPanel{
	private double theta;
	protected Dimension arcs = new Dimension(5, 5);
	
	public IndicatorPanel(){
		super();
		setOpaque(false);
	}
	
	public void setCornerRadius(int rad){
		arcs= new Dimension(rad, rad);
		this.repaint();
	}
	
	public void setRotation(double inTheta){
		//System.out.print("\n"+inTheta);
		theta = inTheta;
		this.repaint();
	}
	public void paintComponent(Graphics g) {
        super.paintComponent(g);

        Graphics2D graphics = (Graphics2D) g;

        int width = getWidth();
        int height = getHeight();
        
        graphics.setRenderingHint(RenderingHints.KEY_ANTIALIASING, 
    			RenderingHints.VALUE_ANTIALIAS_ON);

        //Draws the rounded opaque panel with borders.
        graphics.setColor(getBackground());
        graphics.translate(centerPoint.x-Math.cos(theta)*(indicatorThickness/2), centerPoint.y-Math.sin(theta)*(indicatorThickness/2));
        graphics.rotate(theta);
        graphics.fillRoundRect(0, 0, (int)indicatorThickness, (int)indicatorDim,  arcs.width*2, arcs.height*2);
        //graphics.setColor(getForeground());

    }

}
	

}
